widget/calendar.go

```diff
@@ -0,0 +1,213 @@
+func NewCalendar(cT time.Time, changed func(time.Time)) *Calendar {
+	c := &Calendar{
+		currentTime: cT,
+		OnChanged:   changed,
+	}
+
+	c.ExtendBaseWidget(c)
+
+	return c
+}
+
+// CreateRenderer returns a new WidgetRenderer for this widget.
+// This should not be called by regular code, it is used internally to render a widget.
+func (c *Calendar) CreateRenderer() fyne.WidgetRenderer {
+	c.monthPrevious = NewButtonWithIcon("", theme.NavigateBackIcon(), func() {
+		c.currentTime = c.currentTime.AddDate(0, -1, 0)
+		// Dates are 'normalised', forcing date to start from the start of the month ensures move from March to February
+		c.currentTime = time.Date(c.currentTime.Year(), c.currentTime.Month(), 1, 0, 0, 0, 0, c.currentTime.Location())
+		c.monthLabel.SetText(c.monthYear())
+		c.dates.Objects = c.calendarObjects()
+	})
+	c.monthPrevious.Importance = LowImportance
+
+	c.monthNext = NewButtonWithIcon("", theme.NavigateNextIcon(), func() {
+		c.currentTime = c.currentTime.AddDate(0, 1, 0)
+		c.monthLabel.SetText(c.monthYear())
+		c.dates.Objects = c.calendarObjects()
+	})
+	c.monthNext.Importance = LowImportance
+
+	c.monthLabel = NewLabel(c.monthYear())
+
+	nav := &fyne.Container{Layout: layout.NewBorderLayout(nil, nil, c.monthPrevious, c.monthNext),
+		Objects: []fyne.CanvasObject{c.monthPrevious, c.monthNext,
+			&fyne.Container{Layout: layout.NewCenterLayout(), Objects: []fyne.CanvasObject{c.monthLabel}}}}
+
+	c.dates = &fyne.Container{Layout: newCalendarLayout(), Objects: c.calendarObjects()}
+
+	dateContainer := &fyne.Container{Layout: layout.NewBorderLayout(nav, nil, nil, nil),
+		Objects: []fyne.CanvasObject{nav, c.dates}}
+
+	return NewSimpleRenderer(dateContainer)
+}
+
+func (c *Calendar) calendarObjects() []fyne.CanvasObject {
+	var columnHeadings []fyne.CanvasObject
+	for i := 0; i < daysPerWeek; i++ {
+		j := i + 1
+		if j == daysPerWeek {
+			j = 0
+		}
+
+		t := NewLabel(strings.ToUpper(time.Weekday(j).String()[:3]))
+		t.Alignment = fyne.TextAlignCenter
+		columnHeadings = append(columnHeadings, t)
+	}
+	columnHeadings = append(columnHeadings, c.daysOfMonth()...)
+
+	return columnHeadings
+}
+
+func (c *Calendar) dateForButton(dayNum int) time.Time {
+	oldName, off := c.currentTime.Zone()
+	return time.Date(c.currentTime.Year(), c.currentTime.Month(), dayNum, c.currentTime.Hour(), c.currentTime.Minute(), 0, 0, time.FixedZone(oldName, off)).In(c.currentTime.Location())
+}
+
+func (c *Calendar) daysOfMonth() []fyne.CanvasObject {
+	start := time.Date(c.currentTime.Year(), c.currentTime.Month(), 1, 0, 0, 0, 0, c.currentTime.Location())
+	var buttons []fyne.CanvasObject
+
+	//account for Go time pkg starting on sunday at index 0
+	dayIndex := int(start.Weekday())
+	if dayIndex == 0 {
+		dayIndex += daysPerWeek
+	}
+
+	//add spacers if week doesn't start on Monday
+	for i := 0; i < dayIndex-1; i++ {
+		buttons = append(buttons, layout.NewSpacer())
+	}
+
+	for d := start; d.Month() == start.Month(); d = d.AddDate(0, 0, 1) {
+
```

Comment:
blank lines here seem kind of random and not following an existing grouping style

Reply:
It was ported in from fyne-x. Should we add a new commit just to remove strange empty lines?

Reply:
I thought they can just be cleaned up alongside any commit for any other reason